torchx/lib/torchx.ex

```diff
@@ -1,7 +1,84 @@
+defmodule Torchx.Macro do
+  defmacro __using__(_opts) do
+    quote do
+      import unquote(__MODULE__)
+      Module.register_attribute(Torchx, :torch_function, accumulate: true)
+    end
+  end
+
+  defmacro deftorch(call) do
+    {name, args} = Macro.decompose_call(call)
+
+    args_names = Enum.map(args, &elem(&1, 0))
+
+    if Enum.any?(args, &match?({:device, _, nil}, &1)) do
+      quote do
+        @torch_function {unquote(name), unquote(args_names)}
+        def unquote(name)(unquote_splicing(args)) do
+          Torchx.NIF.call(unquote(name), var!(device), [unquote_splicing(args)])
+          |> wrap(var!(device))
+        end
+      end
+    else
+      quote do
+        @torch_function {unquote(name), unquote(args_names)}
+        def unquote(name)(unquote_splicing(args)) do
+          {device, prepared_args} = prepare_args(unquote(args))
+
+          Torchx.NIF.call(unquote(name), device, prepared_args)
+          |> wrap(device)
+        end
+      end
+    end
+  end
+
+  defmacro deftorch(op, _a, _b) do
+    quote do
+      @torch_function {var!(op), [:tensorA, :tensorB]}
+      def unquote(op)(tensorA, tensorB) do
+        {device, prepared_args} = prepare_args([tensorA, tensorB])
+
+        Torchx.NIF.call(unquote(op), device, prepared_args)
+        |> wrap(device)
+      end
+    end
+  end
+
+  defmacro deftorch(op, _t) do
+    quote do
+      @torch_function {var!(op), [:tensor]}
+      def unquote(op)(tensor) do
+        {device, prepared_args} = prepare_args([tensor])
+
+        Torchx.NIF.call(unquote(op), device, prepared_args)
```

Comment:
Instead of calling `Torchx.NIF.call` and have that call `apply`, I would select the proper IO/CPU operation here, which has the benefit that we can build both ops at compilation time. Something like this:

```elixir
    op_io = :"#{op}_io"

    quote do
      @torch_function {var!(op), [:tensor]}
      def unquote(op)(tensor) do
        case prepare_args([tensor]) do
          {:cpu, prepared_args} -> Torchx.NIF.unquote(op)(unquote_splicing(prepared_args))
          {:io, prepared_args} -> Torchx.NIF.unquote(op_io)(unquote_splicing(prepared_args))
        end
      end
    end
```

It will be slightly faster too. :)